File: agents/08-research-scout/scout.py
"""
Agent 08 — The Research Scout
Surfaces creator economy intelligence from web, LinkedIn, Instagram, and TikTok.
Writes scored, synthesized signals to the Coda Research Intelligence table.
"""

import os, sys, json, time, re, hashlib
import urllib.request, urllib.error
from datetime import datetime, timezone, timedelta
from dotenv import load_dotenv

load_dotenv()

# ── Config ────────────────────────────────────────────────────────────────────
CODA_API_KEY        = os.getenv("CODA_API_KEY")
CODA_DOC_ID         = os.getenv("CODA_DOC_ID", "ktMUNdlobR")
PERPLEXITY_API_KEY  = os.getenv("PERPLEXITY_API_KEY")
APIFY_API_KEY       = os.getenv("APIFY_API_KEY")

RESEARCH_TABLE_ID   = os.getenv("CODA_RESEARCH_INTEL_TABLE_ID", "")  # set after table creation
ANTHROPIC_API_KEY   = os.getenv("ANTHROPIC_API_KEY")

PERPLEXITY_URL = "https://api.perplexity.ai/chat/completions"
APIFY_BASE     = "https://api.apify.com/v2"

DAILY_LIMIT_PER_PLATFORM = 50
WEEKLY_LIMIT_PER_PLATFORM = 150
PERPLEXITY_QUERY_LIMIT = 20

# Load research config
with open(os.path.join(os.path.dirname(__file__), "../../config/research-topics.json")) as f:
    RESEARCH_CONFIG = json.load(f)

with open(os.path.join(os.path.dirname(__file__), "../../config/coda-schema.json")) as f:
    SCHEMA = json.load(f)


# ── HTTP helpers ───────────────────────────────────────────────────────────────
def http_post(url, headers, data):
    body = json.dumps(data).encode()
    req = urllib.request.Request(url, data=body, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            return json.loads(r.read())
    except urllib.error.HTTPError as e:
        raise ValueError(f"HTTP {e.code}: {e.read().decode()[:300]}")


def http_get(url, headers=None):
    req = urllib.request.Request(url, headers=headers or {})
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            return json.loads(r.read())
    except urllib.error.HTTPError as e:
        raise ValueError(f"HTTP {e.code}: {e.read().decode()[:300]}")


# ── Coda helpers ───────────────────────────────────────────────────────────────
CODA_HEADERS = {
    "Authorization": f"Bearer {CODA_API_KEY}",
    "Content-Type": "application/json"
}

def coda_get_recent_urls(days=14):
    """Fetch URLs already in Research Intelligence table (last N days)."""
    if not RESEARCH_TABLE_ID:
        return set()
    cutoff = (datetime.now(timezone.utc) - timedelta(days=days)).isoformat()
    url = f"https://coda.io/apis/v1/docs/{CODA_DOC_ID}/tables/{RESEARCH_TABLE_ID}/rows?limit=500"
    req = urllib.request.Request(url, headers=CODA_HEADERS)
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            data = json.loads(r.read())
        return {row.get("values", {}).get("c-ri-source-url", {}).get("value", "")
                for row in data.get("items", [])}
    except Exception:
        return set()


def coda_write_item(item):
    """Write a single research intelligence item to Coda."""
    if not RESEARCH_TABLE_ID:
        print(f"  [no table ID] Would write: {item['title'][:60]}")
        return
    payload = {"rows": [{"cells": [
        {"column": "c-ri-title",          "value": item["title"]},
        {"column": "c-ri-source-url",     "value": item["url"]},
        {"column": "c-ri-source-type",    "value": item["source_type"]},
        {"column": "c-ri-platform",       "value": item["platform"]},
        {"column": "c-ri-summary",        "value": item["summary"]},
        {"column": "c-ri-raw-excerpt",    "value": item.get("raw_excerpt", "")[:500]},
        {"column": "c-ri-relevance",      "value": item["scores"]["relevance"]},
        {"column": "c-ri-novelty",        "value": item["scores"]["novelty"]},
        {"column": "c-ri-actionability",  "value": item["scores"]["actionability"]},
        {"column": "c-ri-final-score",    "value": item["scores"]["final"]},
        {"column": "c-ri-use-case-tags",  "value": ", ".join(item.get("use_case_tags", []))},
        {"column": "c-ri-priority",       "value": item["scores"]["final"] >= 8.5},
        {"column": "c-ri-date-scouted",   "value": datetime.now(timezone.utc).date().isoformat()},
    ]}]}
    http_post(
        f"https://coda.io/apis/v1/docs/{CODA_DOC_ID}/tables/{RESEARCH_TABLE_ID}/rows",
        CODA_HEADERS, payload
    )


# ── Perplexity web search ──────────────────────────────────────────────────────
def perplexity_search(query, recency="week"):
    """Query Perplexity Sonar for web/news results with citations."""
    headers = {
        "Authorization": f"Bearer {PERPLEXITY_API_KEY}",
        "Content-Type": "application/json"
    }
    data = {
        "model": "sonar",
        "messages": [
            {"role": "system", "content": (
                "You are a research assistant for Cre8te Studio, a community platform "
                "for creators and creative entrepreneurs. Return factual, cited summaries "
                "of recent developments. Focus on specifics: company names, dates, numbers, "
                "direct quotes from announcements. Do not editorialize."
            )},
            {"role": "user", "content": query}
        ],
        "search_recency_filter": recency,
        "return_citations": True,
        "max_tokens": 800
    }
    result = http_post(PERPLEXITY_URL, headers, data)
    content = result["choices"][0]["message"]["content"]
    citations = result.get("citations", [])
    return content, citations


# ── Apify social scrapers ──────────────────────────────────────────────────────
def apify_run_actor(actor_id, input_data, timeout_secs=120):
    """Run an Apify actor and return results."""
    headers = {
        "Authorization": f"Bearer {APIFY_API_KEY}",
        "Content-Type": "application/json"
    }
    # Start run
    run = http_post(
        f"{APIFY_BASE}/acts/{actor_id}/runs?timeout={timeout_secs}",
        headers, {"input": input_data}
    )
    run_id = run.get("data", {}).get("id")
    if not run_id:
        raise ValueError(f"Apify actor {actor_id} failed to start: {run}")

    # Poll for completion
    start = time.time()
    while time.time() - start < timeout_secs + 30:
        status_data = http_get(f"{APIFY_BASE}/actor-runs/{run_id}", headers)
        status = status_data.get("data", {}).get("status", "")
        if status == "SUCCEEDED":
            break
        elif status in ("FAILED", "ABORTED", "TIMED-OUT"):
            raise ValueError(f"Apify run {run_id} ended with status: {status}")
        time.sleep(5)

    # Fetch results
    dataset_id = status_data["data"]["defaultDatasetId"]
    results = http_get(f"{APIFY_BASE}/datasets/{dataset_id}/items?limit=200", headers)
    return results.get("data", {}).get("items", [])


def scrape_linkedin_posts(keywords, limit=50):
    """Pull LinkedIn posts matching keywords via Apify."""
    items = []
    for keyword in keywords[:3]:  # max 3 keywords per run
        try:
            results = apify_run_actor(
                "curious_coder/linkedin-post-search-scraper",
                {"searchQuery": keyword, "maxPosts": min(limit // 3, 20)}
            )
            for r in results:
                items.append({
                    "title": r.get("title", r.get("text", "")[:100]),
                    "url": r.get("postUrl", r.get("url", "")),
                    "platform": "LinkedIn",
                    "source_type": "Social",
                    "raw_excerpt": r.get("text", "")[:500],
                    "engagement": r.get("totalReactionCount", 0) + r.get("commentsCount", 0)
                })
        except Exception as e:
            print(f"  LinkedIn scrape error for '{keyword}': {e}")
    return items[:limit]


def scrape_instagram_hashtags(hashtags, limit=50):
    """Pull Instagram posts from tracked hashtags via Apify."""
    items = []
    for tag in hashtags[:3]:
        try:
            results = apify_run_actor(
                "apify/instagram-hashtag-scraper",
                {"hashtags": [tag], "resultsLimit": min(limit // 3, 20)}
            )
            for r in results:
                caption = r.get("caption", "")
                items.append({
                    "title": caption[:100] if caption else f"#{tag} post",
                    "url": r.get("url", r.get("shortCode", "")),
                    "platform": "Instagram",
                    "source_type": "Social",
                    "raw_excerpt": caption[:500],
                    "engagement": r.get("likesCount", 0) + r.get("commentsCount", 0)
                })
        except Exception as e:
            print(f"  Instagram scrape error for '#{tag}': {e}")
    return items[:limit]


def scrape_tiktok_hashtags(hashtags, limit=50):
    """Pull TikTok posts from tracked hashtags via Apify."""
    items = []
    for tag in hashtags[:3]:
        try:
            results = apify_run_actor(
                "clockworks/tiktok-scraper",
                {"hashtags": [tag], "resultsPerPage": min(limit // 3, 20)}
            )
            for r in results:
                desc = r.get("text", r.get("description", ""))
                items.append({
                    "title": desc[:100] if desc else f"#{tag} video",
                    "url": r.get("webVideoUrl", r.get("videoUrl", "")),
                    "platform": "TikTok",
                    "source_type": "Social",
                    "raw_excerpt": desc[:500],
                    "engagement": r.get("diggCount", 0) + r.get("commentCount", 0) + r.get("shareCount", 0)
                })
        except Exception as e:
            print(f"  TikTok scrape error for '#{tag}': {e}")
    return items[:limit]


# ── Claude synthesis and scoring ──────────────────────────────────────────────
def synthesize_and_score(items, existing_urls):
    """Run Claude scoring pass on raw items. Returns only items above threshold."""
    if not items:
        return []

    # Dedup by URL first
    seen_urls = set(existing_urls)
    deduped = []
    for item in items:
        url = item.get("url", "")
        if url and url not in seen_urls:
            seen_urls.add(url)
            deduped.append(item)

    if not deduped:
        return []

    # Build scoring prompt
    items_text = "\n\n".join([
        f"ITEM {i+1}:\nTitle: {item['title']}\nPlatform: {item['platform']}\n"
        f"Excerpt: {item.get('raw_excerpt','')[:300]}"
        for i, item in enumerate(deduped[:20])  # max 20 per Claude call
    ])

    prompt = f"""You are scoring research items for Cre8te Studio — a community platform 
for creative entrepreneurs, content creators, and personal brand builders.

Score each item 1-10 on three dimensions:
- Relevance: How directly does this relate to creators, creative entrepreneurs, content creation, AI tools for creators, or platform changes affecting creators?
- Novelty: Is this genuinely new information (high) or well-known background (low)?
- Actionability: Does this suggest a specific content angle, workshop topic, newsletter story, or platform update worth covering?

Also assign use case tags from: [Content Idea, Newsletter Story, Workshop Signal, Platform Update, AI Tool, Industry News]
And write a 2-sentence summary from Cre8te Studio's community-first perspective.

Return JSON array only, no other text:
[{{"item_number": 1, "relevance": 8, "novelty": 7, "actionability": 9, "use_case_tags": ["Content Idea", "Newsletter Story"], "summary": "..."}}]

ITEMS TO SCORE:
{items_text}"""

    headers = {
        "x-api-key": ANTHROPIC_API_KEY,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json"
    }
    data = {
        "model": "claude-sonnet-4-20250514",
        "max_tokens": 2000,
        "messages": [{"role": "user", "content": prompt}]
    }

    try:
        result = http_post("https://api.anthropic.com/v1/messages", headers, data)
        raw = result["content"][0]["text"].strip()
        # Strip markdown fences if present
        raw = re.sub(r"^```json\s*", "", raw)
        raw = re.sub(r"\s*```$", "", raw)
        scores_list = json.loads(raw)
    except Exception as e:
        print(f"  Claude scoring error: {e}")
        return []

    scored = []
    for score_data in scores_list:
        idx = score_data.get("item_number", 1) - 1
        if idx >= len(deduped):
            continue
        item = deduped[idx].copy()
        r = score_data.get("relevance", 5)
        n = score_data.get("novelty", 5)
        a = score_data.get("actionability", 5)
        final = round(r * 0.4 + n * 0.3 + a * 0.3, 1)
        item["scores"] = {"relevance": r, "novelty": n, "actionability": a, "final": final}
        item["summary"] = score_data.get("summary", "")
        item["use_case_tags"] = score_data.get("use_case_tags", [])
        if final >= 7.0:
            scored.append(item)

    return sorted(scored, key=lambda x: x["scores"]["final"], reverse=True)


# ── Web search pass ────────────────────────────────────────────────────────────
def run_web_research(query_limit=10):
    """Run Perplexity queries and return formatted items."""
    queries = RESEARCH_CONFIG.get("perplexity_queries", [])[:query_limit]
    items = []
    for query in queries:
        try:
            content, citations = perplexity_search(query)
            for cite_url in citations[:2]:  # top 2 citations per query
                items.append({
                    "title": query,
                    "url": cite_url,
                    "platform": "Web/News",
                    "source_type": "Web",
                    "raw_excerpt": content[:500],
                    "engagement": 0
                })
            time.sleep(1)  # rate limit
        except Exception as e:
            print(f"  Perplexity error for '{query[:50]}': {e}")
    return items


# ── Main ───────────────────────────────────────────────────────────────────────
def main(mode="daily"):
    print(f"\n[Research Scout] Starting {mode} run at {datetime.now(timezone.utc).isoformat()}")

    limit = WEEKLY_LIMIT_PER_PLATFORM if mode == "deep" else DAILY_LIMIT_PER_PLATFORM
    query_limit = PERPLEXITY_QUERY_LIMIT if mode == "deep" else 10

    # Load tracked topics
    li_keywords  = RESEARCH_CONFIG.get("linkedin_keywords", [])
    ig_hashtags  = RESEARCH_CONFIG.get("instagram_hashtags", [])
    tt_hashtags  = RESEARCH_CONFIG.get("tiktok_hashtags", [])

    # Get existing URLs for dedup
    existing_urls = coda_get_recent_urls(days=14)
    print(f"[Research Scout] {len(existing_urls)} items already in Coda (14-day window)")

    # Collect raw items from all sources
    all_items = []

    print("[Research Scout] Running web research (Perplexity)...")
    all_items += run_web_research(query_limit)

    print("[Research Scout] Scraping LinkedIn...")
    all_items += scrape_linkedin_posts(li_keywords, limit)

    print("[Research Scout] Scraping Instagram...")
    all_items += scrape_instagram_hashtags(ig_hashtags, limit)

    print("[Research Scout] Scraping TikTok...")
    all_items += scrape_tiktok_hashtags(tt_hashtags, limit)

    print(f"[Research Scout] {len(all_items)} raw items collected, running Claude scoring...")

    # Score and filter
    scored_items = synthesize_and_score(all_items, existing_urls)
    print(f"[Research Scout] {len(scored_items)} items passed scoring threshold (≥7.0)")

    # Write to Coda
    written, errors = 0, 0
    for item in scored_items:
        try:
            coda_write_item(item)
            priority = "🔴 PRIORITY" if item["scores"]["final"] >= 8.5 else ""
            print(f"  ✓ [{item['scores']['final']}] {item['title'][:60]} {priority}")
            written += 1
            time.sleep(0.3)  # gentle Coda rate limiting
        except Exception as e:
            errors += 1
            print(f"  ERROR writing item: {e}")

    print(f"\n[Research Scout] Done — {written} items written, {errors} errors")

    # Priority alert
    priority_items = [i for i in scored_items if i["scores"]["final"] >= 8.5]
    if priority_items:
        print(f"\n🔴 {len(priority_items)} PRIORITY ITEMS flagged:")
        for item in priority_items:
            print(f"   [{item['scores']['final']}] {item['title'][:80]}")
            print(f"   Tags: {', '.join(item.get('use_case_tags', []))}")
            print(f"   {item.get('summary', '')[:120]}")


if __name__ == "__main__":
    mode = "deep" if "--mode=deep" in sys.argv or "--deep" in sys.argv else "daily"
    main(mode=mode)

File: agents/08-research-scout/scout.ts
/**
 * Agent 08 — The Research Scout
 * Surfaces creator economy intelligence from web, LinkedIn, Instagram, and TikTok.
 * Scores items and writes to Coda Research Intelligence table.
 */
import 'dotenv/config';
import { addRows, sleep, type CodaCell } from '../../lib/coda.js';
import { claudeComplete, parseJsonResponse } from '../../lib/anthropic.js';
import { runActor, type ApifyItem } from '../../lib/apify.js';
import researchTopics from '../../config/research-topics.json' with { type: 'json' };
import schema from '../../config/coda-schema.json' with { type: 'json' };

const RI_TABLE = schema.tables.research_intelligence.table_id;
const RI_COLS  = schema.tables.research_intelligence.columns;

// ── Types ─────────────────────────────────────────────────────────────────────
interface RawItem {
  title:       string;
  url:         string;
  platform:    'LinkedIn' | 'Instagram' | 'TikTok' | 'Web/News';
  source_type: 'Social' | 'Web';
  raw_excerpt: string;
  engagement:  number;
}

interface ScoredItem extends RawItem {
  scores:        { relevance: number; novelty: number; actionability: number; final: number };
  summary:       string;
  use_case_tags: string[];
}

interface ClaudeScore {
  item_number:    number;
  relevance:      number;
  novelty:        number;
  actionability:  number;
  use_case_tags:  string[];
  summary:        string;
}

const DAILY_LIMIT  = 50;
const WEEKLY_LIMIT = 150;
const isDeep       = process.argv.includes('--deep');
const LIMIT        = isDeep ? WEEKLY_LIMIT : DAILY_LIMIT;

// ── Perplexity ────────────────────────────────────────────────────────────────
const PERPLEXITY_KEY = process.env.PERPLEXITY_API_KEY!;

async function perplexitySearch(query: string): Promise<{ content: string; citations: string[] }> {
  const res = await fetch('https://api.perplexity.ai/chat/completions', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${PERPLEXITY_KEY}`,
      'Content-Type':  'application/json',
    },
    body: JSON.stringify({
      model:                 'sonar',
      search_recency_filter: 'week',
      return_citations:       true,
      max_tokens:             800,
      messages: [
        {
          role: 'system',
          content: 'You are a research assistant for Cre8te Studio. Return factual, cited summaries of recent creator economy developments. Be specific: company names, dates, numbers, direct quotes from announcements.',
        },
        { role: 'user', content: query },
      ],
    }),
  });
  if (!res.ok) throw new Error(`Perplexity ${res.status}: ${await res.text()}`);
  const data = await res.json() as {
    choices: Array<{ message: { content: string } }>;
    citations?: string[];
  };
  return {
    content:   data.choices[0].message.content,
    citations: data.citations ?? [],
  };
}

async function runWebResearch(queryLimit: number): Promise<RawItem[]> {
  // Support both old flat list and new categorised object
  const queryConfig = researchTopics.perplexity_queries as
    string[] | { creator_economy: string[]; creative_economy: string[] };

  type QuerySet = { queries: string[]; category: 'creator_economy' | 'creative_economy' };
  const sets: QuerySet[] = Array.isArray(queryConfig)
    ? [{ queries: queryConfig, category: 'creator_economy' }]
    : [
        { queries: queryConfig.creator_economy ?? [], category: 'creator_economy'  },
        { queries: queryConfig.creative_economy ?? [], category: 'creative_economy' },
      ];

  const items: RawItem[] = [];

  for (const { queries, category } of sets) {
    const limited = queries.slice(0, Math.ceil(queryLimit / sets.length));
    for (const query of limited) {
      try {
        const { content, citations } = await perplexitySearch(query);
        for (const url of citations.slice(0, 2)) {
          items.push({
            title:       query,
            url,
            platform:    'Web/News',
            source_type: 'Web',
            raw_excerpt: content.slice(0, 500),
            engagement:  0,
            _category:   category,  // carry through for tagging
          });
        }
        await sleep(1000);
      } catch (err) {
        console.warn(`  Perplexity error [${query.slice(0, 40)}]:`, err);
      }
    }
  }
  return items;
}

// ── Apify scrapers ────────────────────────────────────────────────────────────
async function scrapeLinkedIn(keywords: string[], limit: number): Promise<RawItem[]> {
  const items: RawItem[] = [];
  for (const keyword of keywords.slice(0, 3)) {
    try {
      type LIPost = ApifyItem & { title?: string; text?: string; postUrl?: string; url?: string; totalReactionCount?: number; commentsCount?: number };
      const results = await runActor<LIPost>('curious_coder/linkedin-post-search-scraper', {
        searchQuery: keyword,
        maxPosts:    Math.min(Math.floor(limit / 3), 20),
      });
      for (const r of results) {
        items.push({
          title:       String(r.title ?? r.text ?? '').slice(0, 100),
          url:         String(r.postUrl ?? r.url ?? ''),
          platform:    'LinkedIn',
          source_type: 'Social',
          raw_excerpt:  String(r.text ?? '').slice(0, 500),
          engagement:  Number(r.totalReactionCount ?? 0) + Number(r.commentsCount ?? 0),
        });
      }
    } catch (err) {
      console.warn(`  LinkedIn error [${keyword}]:`, err);
    }
  }
  return items.slice(0, limit);
}

async function scrapeInstagram(hashtags: string[], limit: number): Promise<RawItem[]> {
  const items: RawItem[] = [];
  for (const tag of hashtags.slice(0, 3)) {
    try {
      type IGPost = ApifyItem & { caption?: string; url?: string; shortCode?: string; likesCount?: number; commentsCount?: number };
      const results = await runActor<IGPost>('apify/instagram-hashtag-scraper', {
        hashtags:     [tag],
        resultsLimit: Math.min(Math.floor(limit / 3), 20),
      });
      for (const r of results) {
        const caption = String(r.caption ?? '');
        items.push({
          title:       caption.slice(0, 100) || `#${tag} post`,
          url:         String(r.url ?? r.shortCode ?? ''),
          platform:    'Instagram',
          source_type: 'Social',
          raw_excerpt:  caption.slice(0, 500),
          engagement:  Number(r.likesCount ?? 0) + Number(r.commentsCount ?? 0),
        });
      }
    } catch (err) {
      console.warn(`  Instagram error [#${tag}]:`, err);
    }
  }
  return items.slice(0, limit);
}

async function scrapeTikTok(hashtags: string[], limit: number): Promise<RawItem[]> {
  const items: RawItem[] = [];
  for (const tag of hashtags.slice(0, 3)) {
    try {
      type TTPost = ApifyItem & { text?: string; description?: string; webVideoUrl?: string; videoUrl?: string; diggCount?: number; commentCount?: number; shareCount?: number };
      const results = await runActor<TTPost>('clockworks/tiktok-scraper', {
        hashtags:       [tag],
        resultsPerPage: Math.min(Math.floor(limit / 3), 20),
      });
      for (const r of results) {
        const desc = String(r.text ?? r.description ?? '');
        items.push({
          title:       desc.slice(0, 100) || `#${tag} video`,
          url:         String(r.webVideoUrl ?? r.videoUrl ?? ''),
          platform:    'TikTok',
          source_type: 'Social',
          raw_excerpt:  desc.slice(0, 500),
          engagement:  Number(r.diggCount ?? 0) + Number(r.commentCount ?? 0) + Number(r.shareCount ?? 0),
        });
      }
    } catch (err) {
      console.warn(`  TikTok error [#${tag}]:`, err);
    }
  }
  return items.slice(0, limit);
}

// ── Scoring ────────────────────────────────────────────────────────────────────
async function scoreItems(items: RawItem[], existingUrls: Set<string>): Promise<ScoredItem[]> {
  // URL dedup
  const deduped = items.filter(item => item.url && !existingUrls.has(item.url));
  deduped.forEach(i => existingUrls.add(i.url));

  if (deduped.length === 0) return [];

  // Score in batches of 20
  const scored: ScoredItem[] = [];
  const batch = deduped.slice(0, 20);

  const itemsText = batch
    .map((item, i) =>
      `ITEM ${i + 1}:\nTitle: ${item.title}\nPlatform: ${item.platform}\nExcerpt: ${item.raw_excerpt.slice(0, 300)}`
    )
    .join('\n\n');

  const system = `You are scoring research items for Cre8te Studio — a community platform for creative entrepreneurs and content creators.`;
  const user   = `Score each item 1-10 on:
- relevance: how directly does this relate to creators, creative entrepreneurs, content creation, AI tools for creators, or platform changes affecting creators?
- novelty: is this genuinely new information (high) or well-known background (low)?
- actionability: does this suggest a content angle, workshop topic, newsletter story, or platform update to cover?

Also assign use_case_tags from:
- Content Idea, Workshop Signal, AI Tool, Industry News — general purpose
- Newsletter Story, Platform Update — creator economy: platform changes, monetization tools, ecosystem trends
- Creative Economy — music/film/fashion industry news (ONLY use this for items clearly about those industries)
Write a 2-sentence summary from Cre8te Studio's community-first perspective.

Return JSON array only, no other text:
[{"item_number": 1, "relevance": 8, "novelty": 7, "actionability": 9, "use_case_tags": ["Content Idea"], "summary": "..."}]

ITEMS:
${itemsText}`;

  try {
    const raw     = await claudeComplete(system, user, 2000);
    const results = parseJsonResponse<ClaudeScore[]>(raw);

    for (const score of results) {
      const idx = score.item_number - 1;
      if (idx >= batch.length) continue;
      const item  = batch[idx];
      const r     = score.relevance;
      const n     = score.novelty;
      const a     = score.actionability;
      const final = Math.round((r * 0.4 + n * 0.3 + a * 0.3) * 10) / 10;
      if (final >= 7.0) {
        scored.push({ ...item, scores: { relevance: r, novelty: n, actionability: a, final }, summary: score.summary, use_case_tags: score.use_case_tags });
      }
    }
  } catch (err) {
    console.warn('  Claude scoring error:', err);
  }

  return scored.sort((a, b) => b.scores.final - a.scores.final);
}

// ── Coda write ─────────────────────────────────────────────────────────────────
async function writeItem(item: ScoredItem): Promise<void> {
  const cells: CodaCell[] = [
    { column: RI_COLS.item_title,      value: item.title },
    { column: RI_COLS.source_url,      value: item.url },
    { column: RI_COLS.source_type,     value: item.source_type },
    { column: RI_COLS.platform,        value: item.platform },
    { column: RI_COLS.summary,         value: item.summary },
    { column: RI_COLS.raw_excerpt,     value: item.raw_excerpt.slice(0, 500) },
    { column: RI_COLS.relevance_score, value: item.scores.relevance },
    { column: RI_COLS.novelty_score,   value: item.scores.novelty },
    { column: RI_COLS.actionability,   value: item.scores.actionability },
    { column: RI_COLS.final_score,     value: item.scores.final },
    { column: RI_COLS.use_case_tags,   value: item.use_case_tags.join(', ') },
    { column: RI_COLS.priority_flag,   value: item.scores.final >= 8.5 },
    { column: RI_COLS.date_scouted,    value: new Date().toISOString().split('T')[0] },
  ];
  await addRows(RI_TABLE, [cells]);
}

// ── Main ───────────────────────────────────────────────────────────────────────
async function main(): Promise<void> {
  console.log(`\n[Research Scout] Starting ${isDeep ? 'DEEP' : 'daily'} run at ${new Date().toISOString()}`);

  const topics = researchTopics as {
    linkedin_keywords:  string[];
    instagram_hashtags: string[];
    tiktok_hashtags:    string[];
    perplexity_queries: string[];
  };

  const queryLimit = isDeep ? 20 : 10;
  const existingUrls = new Set<string>(); // populated from Coda in production

  const allItems: RawItem[] = [];

  console.log('[Research Scout] Web research (Perplexity)...');
  allItems.push(...await runWebResearch(queryLimit));

  console.log('[Research Scout] LinkedIn scrape...');
  allItems.push(...await scrapeLinkedIn(topics.linkedin_keywords, LIMIT));

  console.log('[Research Scout] Instagram scrape...');
  allItems.push(...await scrapeInstagram(topics.instagram_hashtags, LIMIT));

  console.log('[Research Scout] TikTok scrape...');
  allItems.push(...await scrapeTikTok(topics.tiktok_hashtags, LIMIT));

  console.log(`[Research Scout] ${allItems.length} raw items — running scoring...`);

  const scored = await scoreItems(allItems, existingUrls);
  console.log(`[Research Scout] ${scored.length} items passed threshold (≥ 7.0)`);

  let written = 0, errors = 0;
  for (const item of scored) {
    try {
      await writeItem(item);
      const flag = item.scores.final >= 8.5 ? ' 🔴 PRIORITY' : '';
      console.log(`  ✓ [${item.scores.final}] ${item.title.slice(0, 60)}${flag}`);
      written++;
      await sleep(300);
    } catch (err) {
      errors++;
      console.error(`  ERROR writing item:`, err);
    }
  }

  console.log(`\n[Research Scout] Done — ${written} written, ${errors} errors`);

  const priority = scored.filter(i => i.scores.final >= 8.5);
  if (priority.length > 0) {
    console.log(`\n🔴 ${priority.length} PRIORITY items:`);
    for (const item of priority) {
      console.log(`  [${item.scores.final}] ${item.title.slice(0, 70)}`);
      console.log(`  Tags: ${item.use_case_tags.join(', ')}`);
      console.log(`  ${item.summary.slice(0, 120)}`);
    }
  }
}

main().catch(err => { console.error(err); process.exit(1); });
